analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ContainsInsteadOfAny.cs

```diff
@@ -0,0 +1,239 @@
+        intList.Any(x => 0 == 0); // Compliant
+
+        intList.Any(x => x.Equals(x)); // Compliant
+        intList.Any(x => someInt.Equals(anotherInt)); // Compliant
+        intList.Any(x => someInt.Equals(0)); // Compliant
+        intList.Any(x => 0.Equals(0)); // Compliant
+        intList.Any(x => x.Equals(x + 1)); // Compliant
+
+        intList.Any(x => x.GetType() == typeof(int)); // Compliant
+        intList.Any(x => x.GetType().Equals(typeof(int))); // Compliant
+        intList.Any(x => MyIntCheck(x)); // Compliant
+        intList.Any(MyIntCheck);  // Compliant
+        intList.Any(x => x != 0); // Compliant
+        intList.Any(x => x.Equals(0) && true);   // Compliant
+        intList.Any(x => (x == 0 ? 2 : 0) == 0); // Compliant
+        intList.Any(x => { return x == 0; });    // Compliant
+
+        stringList.Any(x => x == ""); // Noncompliant
+        stringList.Any(x => "" == x); // Noncompliant
+        stringList.Any(x => x == someString); // Noncompliant
+        stringList.Any(x => someString == x); // Noncompliant
+        stringList.Any(x => x.Equals("")); // Noncompliant
+        stringList.Any(x => "".Equals(x)); // Noncompliant
+        stringList.Any(x => Equals(x, "")); // Noncompliant
+        stringList.Any(x => x is ""); // Compliant FN
+        stringList.Any(x => "" is x); // Error [CS0150]
+        stringList.Any(x => x is someString); // Error [CS0150]
+        stringList.Any(x => someString is x); // Error [CS0150]
+
+        stringList.Any(x => MyStringCheck(x)); // Compliant
+        stringList.Any(MyStringCheck);  // Compliant
+        stringList.Any(x => x != "");   // Compliant
+        stringList.Any(x => x.Equals("") && true);   // Compliant
+        stringList.Any(x => (x == "" ? "a" : "b") == "a"); // Compliant
+        stringList.Any(x => x.Equals("" + someString)); // Compliant
+
+        intArray.Any(x => x == 0); // Compliant
+        intArray.Any(x => 0 == x); // Compliant
+        intArray.Any(x => x == someInt); // Compliant
+        intArray.Any(x => someInt == x); // Compliant
+        intArray.Any(x => x.Equals(0));  // Compliant
+        intArray.Any(x => 0.Equals(x));  // Compliant
+        intArray.Any(x => someInt.Equals(x)); // Compliant
+        intArray.Any(x => x.Equals(x + 1));   // Compliant
+
+        refList.Any(x => x == someRef); // Compliant
+        refList.Any(x => someRef == x); // Compliant
+        refList.Any(x => x.Equals(someRef));  // Noncompliant
+        refList.Any(x => someRef.Equals(x));  // Noncompliant
+        refList.Any(x => Equals(someRef, x)); // Noncompliant
+        refList.Any(x => Equals(x, someRef)); // Noncompliant
+        refList.Any(x => x is someRef); // Error [CS0150]
+        refList.Any(x => someRef is x); // Error [CS0150]
+
+        intList.Any(x => x == null); // Noncompliant FP (warning: the result of this expression will always be false since a value-type is never equal to null)
```

Comment:
It's not that easy, unfortunately. We should both check if one of the two parts is a `NullLiteralExpression` and then check the type of the other part, which is the lambda variable. To do that you cannot just go to the left side of the arrow and ask the type because it's not a declaration, so you need to go to the list itself and pick the type from there. It's quite messy and this rule is already not that easy to read, moreover, those cases are already flagged by a warning (it's weird that it compiles a null check on a value-type in the first place). My 5 cents: if you are writing this line in your code this FP is the least of your problems  :sweat_smile: